## About
**SOS** is a pipeline for *de novo* transcriptome assembly of one or more RNA-seq datasets.
It currently features:
* error correction with [SEECER](http://sb.cs.cmu.edu/seecer/)
* Read reduction [ORNA](https://github.com/SchulzLab/ORNA)
* de novo assembly
* transcript quantification using [Salmon](https://github.com/COMBINE-lab/salmon)
* automatic parameter estimation of kmer values for the assembly and error correction [KREATION](https://github.com/SchulzLab/KREATION
* config file support to streamline many datasets

##### Contact
For questions or suggestions regarding SOS please contact

* Dilip A Durai (ddurai_at_contact.mmci.uni-saarland.de)
* Marcel H Schulz (mschulz_at_mmci.uni-saarland.de)

##### Version
Version 0.2

## Requirements
for SOS:

1.	64-bit linux operating system. 
2.	A physical memory of 12 GB or higher is recommended (more is always better!)
3.	[Jellyfish](http://www.cbcb.umd.edu/software/jellyfish/) version 2.0  
4.	GNU Scientific Library (SEECER)
5.  Any version of [g++](gcc.gnu.org) >= 4.7 (ORNA and Salmon)


## Download
The pipeline can be downloaded using the following command:

    git clone https://github.com/SchulzLab/SOS

The downloaded folder should have the following files:

* Snakemake file
* Sample config file for snakemake: config.yaml
* install_script.py	
* export.sh
* A folder consisting of sample data: Sample_Data
* A scripts folder consisting of supporting scripts: scripts

## Installation

1.	After downloading the SOS distribution, change into the directory
		cd SOS

2.	Then run the python script install_script.py to install the softwares (SEECER,OASES and SAILFISH)
	```
		python install_script.py -f <destination_folder>
	```
	This should install SEECER, OASES, ORNA and SALMON in the provided destination folder. Additional assembly algorithms can be downloaded seperately.  
	Note: Running the script is optional. The users can download and install the softwares seperately  

3.	The PATH variable should contain executables of each software and the LD_LIBRARY_PATH variable should contain 'lib' folder of Sailfish.
	To make this process easier we have included a bash script export.sh. To use this script type the following command:
	```
		source export.sh <folder_name>
	```
	Note: To make use of this script all the softwares should be installed in the same directory (folder_name)

## Config file
The config.txt should have the following configuration
```
#General Parameters
input: input_files //Two comma seperated files if input is paired end data
outdir: Output directory to store the results 
kmer: kmer to be used for error correction, normalization and assembly
#Read Characteristics
type: single or paired 
readlength: 101  
inslength: 200
#Seecer Parameters (See seecer manual for more details)
seecertmp: path to temp folder for seecer intermediate files
jellyfish: path to jellyfish
scriptsfolder: path to the scripts folder of seecer
seecerdir: path to the bin folder of seecer
seecerkmer: a seecer kmer if different than the kmer from general parameters
#ORNA parameters
ornabase: Base parameter of ORNA (set to 1.7 by default)
#KREATION Parameters (See KREATION manual for more details)
kstep: step size for kmer parameter of KREATION (set to 2 by default)
kthreshold: Threshold parameter of seecer (set to 0.01 by default)
kpname: Assembly algorithm to be used with KREATION (Current supported assemblers: Oases, TransABySS, TransLiG and SOAPdenovotrans)
kpadditional: Additional commands to be used with KREATION 
```

## Usage


###### Output Folder
The output folder will have a folders generated by the oases run and the folders generated by the sailfish run. 
It will also have a folder named Final_Output which will contain the following files :

1.	reads_corrected.fa (result of SEECER)
2.	transcripts.fa (result of OASES)
3.	quant.sf (result of SAILFISH)

If the read reduction option is used then the output folder will contain folder named *Jellyfish_Output* which stores the intermediate results generated by the read reduction step.
